Development/nmos/control_protocol_methods.cpp

```diff
@@ -0,0 +1,568 @@
+
+            slog::log<slog::severities::more_info>(gate, SLOG_FLF) << "Get property: " << property_id.serialize();
+
+            // find the relevant nc_property_descriptor
+            const auto& property = find_property(parse_nc_property_id(property_id), parse_nc_class_id(nmos::fields::nc::class_id(resource->data)), get_control_protocol_class);
+            if (!property.is_null())
+            {
+                return make_control_protocol_message_response(handle, { nc_method_status::ok }, resource->data.at(nmos::fields::nc::name(property)));
+            }
+
+            // unknown property
+            utility::stringstream_t ss;
+            ss << U("unknown property: ") << property_id.serialize() << U(" to do Get");
+            return make_control_protocol_error_response(handle, { nc_method_status::property_not_implemented }, ss.str());
+        }
+
+        // Set property value
+        web::json::value set(nmos::resources& resources, nmos::resources::iterator resource, int32_t handle, const web::json::value& arguments, get_control_protocol_class_handler get_control_protocol_class, get_control_protocol_datatype_handler, slog::base_gate& gate)
+        {
+            // note, model mutex is already locked by the outter function, so access to control_protocol_resources is OK...
+
+            const auto& property_id = nmos::fields::nc::id(arguments);
+            const auto& val = nmos::fields::nc::value(arguments);
+
+            slog::log<slog::severities::more_info>(gate, SLOG_FLF) << "Set property: " << property_id.serialize() << " value: " << val.serialize();
+
+            // find the relevant nc_property_descriptor
+            const auto& property = find_property(parse_nc_property_id(property_id), parse_nc_class_id(nmos::fields::nc::class_id(resource->data)), get_control_protocol_class);
+            if (!property.is_null())
+            {
+                if (nmos::fields::nc::is_read_only(property))
+                {
+                    return make_control_protocol_message_response(handle, { nc_method_status::read_only });
+                }
+
+                if ((val.is_null() && !nmos::fields::nc::is_nullable(property))
+                    || (val.is_array() && !nmos::fields::nc::is_sequence(property)))
+                {
+                    return make_control_protocol_message_response(handle, { nc_method_status::parameter_error });
+                }
+
+                const auto notification = make_control_protocol_notification(nmos::fields::nc::oid(resource->data), nc_object_property_changed_event_id, { parse_nc_property_id(property_id), nc_property_change_type::type::value_changed, val });
+                const auto notification_event = make_control_protocol_notification(web::json::value_of({ notification }));
+
+                modify_resource(resources, resource->id, [&](nmos::resource& resource)
+                {
+                    resource.data[nmos::fields::nc::name(property)] = val;
```

Comment:
Can this segfault? `val` is use in the constructor of  `nc_property_changed_event_data` (line 63) which uses a std::move for the value. So `val` might actually be undefined later.

Reply:
Thank you @maweit for spotting it, I will correct them all.